### PR TITLE
Support serialization of ByteArrays

### LINKED ISSUE #1330: Add support for ByteArray
[Cairo 2.4.0](https://community.starknet.io/t/cairo-v2-4-0-is-out/109275) adds ByteArrays. Add support for contract with this type.

### PR DESCRIPTION
**Pull request by ddoktorski**
Closes #1330 

## Introduced changes

- Enable serializing and deserializing Cairo `ByteArray` into and from Python `str`
- Allow using contracts that utilize `ByteArray` type,  in contract serialization logic:
  - `bytes_31` is treated as `felt`
  - `ByteArray` is handled as a special case of a struct instead of a separate type

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/serialization/data_serializers/byte_array_serializer_test.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import pytest
+
+from starknet_py.serialization.data_serializers.byte_array_serializer import (
+    ByteArraySerializer,
+)
+
+byte_array_serializer = ByteArraySerializer()
+
+
+@pytest.mark.parametrize(
```
Comment (ddoktorski) [inline starknet_py/serialization/data_serializers/byte_array_serializer_test.py:10]: Note: Examples taken from https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types/#serialization_of_byte_arrays

File: starknet_py/serialization/data_serializers/byte_array_serializer.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+from starknet_py.serialization.data_serializers._common import (
+    deserialize_to_list,
+    serialize_from_list,
+)
+from starknet_py.serialization.data_serializers.cairo_data_serializer import (
+    CairoDataSerializer,
+)
+from starknet_py.serialization.data_serializers.felt_serializer import FeltSerializer
+
+BYTES_SIZE_IN_BYTES_31 = 31
```
[the remark this answers is not in the document]
Comment (ddoktorski) [reply starknet_py/serialization/data_serializers/byte_array_serializer.py:18]: In Cairo it is [BYTES_IN_BYTES_31](https://github.com/starkware-libs/cairo/blob/e4e155d6e82bb06d3da6c98c923ce8a9da03a08d/corelib/src/bytes_31.cairo#L5), maybe rename it to just `BYTES_31_SIZE`?

[the remark this answers is not in the document]
Comment (THenry14) [reply starknet_py/serialization/data_serializers/byte_array_serializer.py:18]: yeah, probably a good idea

File: starknet_py/serialization/data_serializers/byte_array_serializer.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        with context.push_entity("data_array_len"):
+            [size] = context.reader.read(1)
+
+        data = deserialize_to_list([FeltSerializer()] * size, context)
+
+        with context.push_entity("pending_word"):
+            [pending_word] = context.reader.read(1)
+
+        with context.push_entity("pending_word_len"):
+            [pending_word_len] = context.reader.read(1)
```
[the remark this answers is not in the document]
Comment (ddoktorski) [reply starknet_py/serialization/data_serializers/byte_array_serializer.py:41]: It follows the standard [serialization of a struct](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types/#serialization_of_structs). The `pending_word` and `pending_word_len` are just members of the `ByteArray` struct.

[the remark this answers is not in the document]
Comment (THenry14) [reply starknet_py/serialization/data_serializers/byte_array_serializer.py:41]: I think you meant [serialization of byte arrays](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types/#serialization_of_byte_arrays). Indeed docs state the length is the last element, thanks :+1:

**Review by THenry14 — APPROVED** [review]

**Pull request merged**